client/src/components/base-editor/base-editor.tsx

```diff
@@ -24,24 +27,94 @@ const BaseEditor = forwardRef(
     ref: React.Ref<MonacoEditor>,
   ) => {
     const { textEditorSettings } = useContext(SettingsContext)
+    const { customBinding, theme } = textEditorSettings
+
+    const boundEditor = useRef<any>(undefined)
+    const editor = useRef<monaco.editor.IStandaloneCodeEditor | undefined>(
+      undefined,
+    )
+    const actionsContainer = useRef<HTMLDivElement | null>(null)
+
+    useEffect(() => {
+      if (customBinding === 'none' && boundEditor.current) {
+        boundEditor.current.dispose()
+      } else if (
+        customBinding === 'vim' &&
+        editor.current &&
+        actionsContainer.current
+      ) {
+        // if switching directly from emacs
+        if (boundEditor.current) {
+          boundEditor.current.dispose()
+        }
+
+        boundEditor.current = initVimMode(
+          editor.current,
+          actionsContainer.current,
+        )
+      } else if (
+        customBinding === 'emacs' &&
+        editor.current &&
+        actionsContainer.current
+      ) {
+        // if switching directly from vim
+        boundEditor.current && boundEditor.current.dispose()
+
+        boundEditor.current = new EmacsExtension(editor.current)
+        boundEditor.current.onDidMarkChange((ev: any) => {
+          if (actionsContainer.current !== null)
+            actionsContainer.current.textContent = ev
+              ? 'Mark Set!'
+              : 'Mark Unset'
+        })
+        boundEditor.current.onDidChangeKey((str: string) => {
+          if (actionsContainer.current !== null)
+            actionsContainer.current.textContent = str
+        })
+        boundEditor.current.start()
+      }
+    }, [customBinding])
+
+    useEffect(() => {
```

Comment:
We can create a custom hook `useEditorTheme` and use it as `useTheme(textEditorSettings.theme)`. It would make the code more readable and we can abstract away the implementation details to another hook.

```typescript
function useEditorTheme(theme: string) {
  useEffect(() => {
    // update editor theme based on the `theme` input
  }, [theme])
}
```